From 2554b012c0e24dbb4f6051ce4e589f1b66ec62e7 Mon Sep 17 00:00:00 2001
From: Victor Gu <xigu@marvell.com>
Date: Fri, 9 Dec 2016 16:37:58 +0800
Subject: [PATCH 0804/1345] uart: a3700: add initial PM support

commit  7115bcf539a838fba679db0f503e7152be810904 from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

This patch added suspend/resume support for Armada37x0's
primary UART port.
Currently, the UART driver still doesn't support to restore
the configurations, such as baudrate settings and interrupt
mode, etc. And the secondary UART port requires these
additonal configurations.

Change-Id: I8ea8784c3e3b77c8ad98871bc1b0c4d7d248df79
Signed-off-by: Victor Gu <xigu@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/34538
Reviewed-by: Wilson Ding <dingwei@marvell.com>
Tested-by: Wilson Ding <dingwei@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/36506
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/tty/serial/mvebu-uart.c |   34 ++++++++++++++++++++++++++++++++++
 1 file changed, 34 insertions(+)

diff --git a/drivers/tty/serial/mvebu-uart.c b/drivers/tty/serial/mvebu-uart.c
index e3cb718..bd03df2 100644
--- a/drivers/tty/serial/mvebu-uart.c
+++ b/drivers/tty/serial/mvebu-uart.c
@@ -980,6 +980,37 @@ static int mvebu_uart_probe(struct platform_device *pdev)
 	return 0;
 }
 
+#ifdef CONFIG_PM
+static int mvebu_uart_suspend(struct device *dev)
+{
+	struct mvebu_uart_data *data = dev_get_drvdata(dev);
+
+	if (data->port)
+		uart_suspend_port(&mvebu_uart_driver, data->port);
+
+	device_set_wakeup_enable(dev, true);
+
+	return 0;
+}
+
+static int mvebu_uart_resume(struct device *dev)
+{
+	struct mvebu_uart_data *data = dev_get_drvdata(dev);
+
+	if (data->port)
+		uart_resume_port(&mvebu_uart_driver, data->port);
+
+	device_set_wakeup_enable(dev, false);
+
+	return 0;
+}
+
+static const struct dev_pm_ops mvebu_uart_pm_ops = {
+	.suspend	= mvebu_uart_suspend,
+	.resume		= mvebu_uart_resume,
+};
+#endif
+
 /* Match table for of_platform binding */
 static const struct of_device_id mvebu_uart_of_match[] = {
 	{ .compatible = "marvell,armada-3700-uart",     .data = (void *)REG_UART_A3700     },
@@ -992,6 +1023,9 @@ static int mvebu_uart_probe(struct platform_device *pdev)
 	.driver	= {
 		.name  = "mvebu-uart",
 		.of_match_table = of_match_ptr(mvebu_uart_of_match),
+#ifdef CONFIG_PM
+		.pm	= &mvebu_uart_pm_ops,
+#endif
 		.suppress_bind_attrs = true,
 	},
 };
-- 
1.7.9.5

